OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+                                                const
+{
+    ImplicitResidual residualStruct = calcImplicitResidual(s,
+                                                           projFibVelNorm_guess,
+                                                           activdot_guess,
+                                                           u, 0);
+
+    SimTK::Vec2 residual = {residualStruct.forceResidual,
+                            residualStruct.activResidual};
+    return(residual);
+}
+
+
+
+//These are hacks because I am not using muscle.h mli cache variables, yet:
+double  VandenBogert2011Muscle::getFiberLength(SimTK::State& s) const
+{
+    double fiberLength=getStateVariableValue(s, "fiber_length");
+    return fiberLength;
+}
+double  VandenBogert2011Muscle::getActivation(SimTK::State& s) const
+{
+    double activ=getStateVariableValue(s, "activation");
+    return activ;
+}
+
+
+
+//==============================================================================
+// Muscle.h Interface
+//==============================================================================
+
+
+double  VandenBogert2011Muscle::computeActuation(const SimTK::State& s) const
+{
+    double activ=getActivation(s);
+    // For now we will start with static guess
+    // TODO: figure out how to use previous state as guess
+    SimTK::Vector ydotInitialGuess(2);
+    ydotInitialGuess[0]=0.0;
+    ydotInitialGuess[1]=0.0;
+    SimTK::Vec2 sdot=calcSolveMuscle(s, activ, ydotInitialGuess);
+
+    VandenBogert2011Muscle::ImplicitResidual Result= calcImplicitResidual(s,
+                                          sdot[0],
+                                          sdot[1],
+                                          activ,
+                                          0);
+
+    //TODO: Could probably add a check here that residuals =0
+    double actuation=Result.forceTendon;
+
+    return(actuation);
+}
+
+
+
+void VandenBogert2011Muscle::computeInitialFiberEquilibrium(SimTK::State& s)
+const
+{
+    // TODO:  this is the same code as computeFiberEquilibriumAtZeroVelocity
+    //Need to update for v~=0
+    //Let's start with the default activation and the fiberVelocity=0
+
+    computeFiberEquilibriumAtZeroVelocity(s);
+
+}
+
+
+
+void VandenBogert2011Muscle::
+computeFiberEquilibriumAtZeroVelocity(SimTK::State& s) const
+{
+    double activation = getDefaultActivation();
+    SimTK::Vec2 lenghAndForce=calcFiberStaticEquilbirum(getLength(s),
+                                                        activation);
+
+    double projFibLenNorm=lenghAndForce[0];
+    double tendonForce=lenghAndForce[1];
+
+    //TODO:  Millard Muscle handles non-convergence here and zero muscle length.
+    //     Need to consider adding.
+
+    setActuation(s, tendonForce);
+    double fiberLength =
+            projFibLenToFiberLength(projFibLenNorm,1)*getOptimalFiberLength();
+    setStateVariableValue(s, "fiber_length", fiberLength);
+    setStateVariableValue(s, "activation",  activation);
```

Comment:
I believe this method should not be using default activation, and should not set the value of the activation state variable. It should only set fiber length.